### PR TITLE
Fix for Issue #346 : Provisioning hash logging feature

### PR DESCRIPTION
**Pull request by jaylogue**
- Added a method to the Device Layer ConfigurationManager interface to compute a provisioning hash value for the device's provisioning data.  The output of the method is a SHA256 hash value that confirms to the hash algorithm described in the Nest Weave: Factory Provisioning Specification.

- Added code to GenericConfigurationManagerImpl<> to implement the new ComputeProvisioningHash() method.

- Added a Device Layer configuration option (WEAVE_DEVICE_CONFIG_LOG_PROVISIONING_HASH) that causes the device’s provisioning hash to be logged at boot time.

- Fixed a rounding bug in the BASE64_ENCODED_LEN macro that overstated the expected length of base-64 encoded data.

### CODE REVIEW

File: src/adaptations/device-layer/include/Weave/DeviceLayer/internal/GenericConfigurationManagerImpl.ipp [hunk 1]
```diff
@@ -88,8 +89,29 @@ WEAVE_ERROR GenericConfigurationManagerImpl<ImplClass>::_ConfigureWeaveStack()
...
+
+#if WEAVE_DEVICE_CONFIG_LOG_PROVISIONING_HASH
+    {
+        uint8_t provHash[Platform::Security::SHA256::kHashLength];
+        char provHashBase64[BASE64_ENCODED_LEN(sizeof(provHash)) + 1];
+        err = Impl()->_ComputeProvisioningHash(provHash, sizeof(provHash));
+        if (err == WEAVE_NO_ERROR)
+        {
+            Base64Encode(provHash, sizeof(provHash), provHashBase64);
+            provHashBase64[sizeof(provHashBase64) - 1] = 0;
```
Comment (gerickson) [inline src/adaptations/device-layer/include/Weave/DeviceLayer/internal/GenericConfigurationManagerImpl.ipp]: Style: minor issue; however, I find that code that follows "= '\0'" makes it more clear that NULL termination is being performed.

File: src/adaptations/device-layer/include/Weave/DeviceLayer/internal/GenericConfigurationManagerImpl.ipp [hunk 2]
```diff
@@ -693,6 +715,146 @@ bool GenericConfigurationManagerImpl<ImplClass>::_IsPairedToAccount()
...
+    //     CCCC is the length of the cccc…cccc field (the device certificate), represented as 4 uppercase hex digits.
+    //     KKKK is the length of the kkkk…kkkk field (the device private key), represented as 4 uppercase hex digits.
+    //     PPPP is the length of the pppppp field (the device pairing code), represented as 4 uppercase hex digits.
+
+    hash.Begin();
+
+    // Hash the device id
+    {
+        uint64_t deviceId;
+        char inputBuf[21]; // 16 hex character device id + 4 hex character length + terminator
```
Comment (gerickson) [inline src/adaptations/device-layer/include/Weave/DeviceLayer/internal/GenericConfigurationManagerImpl.ipp]: Style: magic number + 1.

File: src/adaptations/device-layer/include/Weave/DeviceLayer/internal/GenericConfigurationManagerImpl.ipp [hunk 3]
```diff
@@ -693,6 +715,146 @@ bool GenericConfigurationManagerImpl<ImplClass>::_IsPairedToAccount()
...
+
+        err = Impl()->_GetDeviceId(deviceId);
+        SuccessOrExit(err);
+
+        snprintf(inputBuf, sizeof(inputBuf), "0010%016" PRIX64, deviceId);
+
+        hash.AddData((uint8_t *)inputBuf, 20);
```
Comment (gerickson) [inline src/adaptations/device-layer/include/Weave/DeviceLayer/internal/GenericConfigurationManagerImpl.ipp]: Style: magic number: 20.

File: src/adaptations/device-layer/include/Weave/DeviceLayer/internal/GenericConfigurationManagerImpl.ipp [hunk 4]
```diff
@@ -693,6 +715,148 @@ bool GenericConfigurationManagerImpl<ImplClass>::_IsPairedToAccount()
...
+        Platform::Security::MemoryFree(dataBuf);
+    }
+    return err;
+}
+
+template<class ImplClass>
+void GenericConfigurationManagerImpl<ImplClass>::HashLengthAndBase64Value(Platform::Security::SHA256 & hash, const uint8_t * val, uint16_t valLen)
+{
+    constexpr uint16_t kInputBufSize = 80;
+    static_assert(kInputBufSize > 0 && kInputBufSize % 4 == 0, "kInputBufSize must be a positive multiple of 4");
```
Comment (gerickson) [inline src/adaptations/device-layer/include/Weave/DeviceLayer/internal/GenericConfigurationManagerImpl.ipp]: For portability, I'd recommend using the nlSTATIC_ASSERT macros in nlassert.

Comment (jaylogue) [reply src/adaptations/device-layer/include/Weave/DeviceLayer/internal/GenericConfigurationManagerImpl.ipp]: I don't believe there's any portability issue here as static_assert in this form is C++11 and C++11 is the minimum language level supported by the Device Layer.

**Pull request merged**